pkg/store/prometheus.go

```diff
@@ -220,6 +225,55 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
 	return p.handleStreamedPrometheusResponse(s, httpResp, queryPrometheusSpan, extLset)
 }
 
+func (p *PrometheusStore) queryPrometheus(s storepb.Store_SeriesServer, r *storepb.SeriesRequest) error {
+	ctx := context.Background()
+	queryOpts := promclient.QueryOptions{
+		Deduplicate:             false,
+		PartialResponseStrategy: storepb.PartialResponseStrategy_WARN,
+		Method:                  "GET",
+	}
+
+	matrix, _, err := p.client.QueryRange(ctx, p.base, r.ToPromQL(), r.MinTime, r.MaxTime, r.QueryHints.Step/1000, queryOpts)
```

Comment:
Wouldn't it be possible to pass down the same hints via remote read and read streamed responses that way: https://github.com/prometheus/prometheus/blob/39d79c3cfb86c47d6bc06a9e9317af582f1833bb/prompb/remote.proto#L30-L70? :thinking:

Reply:
Interesting, I didn't think about this. Would remote read be able to apply the aggregations from the hints?

Reply:
Prometheus passes down the hints to Select() so I think it should work: https://github.com/prometheus/prometheus/blob/main/storage/remote/read_handler.go#L141-L155. Perhaps worth adding a test for testing at least one of these functions that it returns the same result as in: https://github.com/thanos-io/thanos/pull/4712/files#diff-3a68fda5485fac918834a9a7d2c3522f859de6e21f0b41ff2031eebb2579eed0? @bwplotka perhaps I am wrong, feel free to correct me

Reply:
So hints do not do anything right now if you use them with remote read (: 

We have multiple options:

1. Add PromQL pass for remote read hints
2. Optimize Query API to stream series to us

All need to be done on the Prom side to leverage that. I think in mean time we could do it naively using Query API - that's the closest to what we want (: 

I think remote read + hints that pre-aggregates would be nice if community will accept it. I can try to start putting issues about that. WDYT?

Reply:
:+1: sounds good, so let's start with a JSON-based implementation. (1) seems better, it should be relatively easy to take the provided hints into account while executing? :thinking:  :smile: anyway, but that's for the future. But let's add a feature flag, right?